File: examples/v1_3_0_features_demo.py
"""Depyler v1.3.0 Advanced Type System Features Demo

This demonstrates the features implemented so far in v1.3.0
"""

# 1. With Statement Support
class ResourceManager:
    """A simple resource manager for with statements"""
    
    def __init__(self, name: str):
        self.name = name
        self.is_open = False
    
    def __enter__(self):
        """Enter the context"""
        self.is_open = True
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        """Exit the context"""
        self.is_open = False
        return False
    
    def use_resource(self) -> int:
        """Use the resource"""
        if self.is_open:
            return 42
        return 0

def demo_with_statement():
    """Demonstrate with statement support"""
    with ResourceManager("test") as rm:
        result = rm.use_resource()
    return result

# 2. Iterator Protocol Support
class Counter:
    """A simple counter iterator"""
    
    def __init__(self, max_count: int):
        self.max_count = max_count
        self.count = 0
    
    def __iter__(self):
        """Return self as iterator"""
        return self
    
    def __next__(self) -> int:
        """Get next value"""
        if self.count < self.max_count:
            self.count += 1
            return self.count
        return -1  # Simplified StopIteration

def demo_iterator():
    """Demonstrate iterator protocol"""
    counter = Counter(3)
    
    # Manual iteration (for loops with custom iterators need more work)
    total = 0
    val = counter.__next__()
    while val != -1:
        total += val
        val = counter.__next__()
    
    return total

# Main demo function
def main():
    """Run all v1.3.0 feature demos"""
    with_result = demo_with_statement()
    iter_result = demo_iterator()
    
    return with_result + iter_result

File: examples/v1_3_0_features_demo.rs
use serde_json;
#[derive(Debug, Clone)]
pub struct ResourceManager {
    pub name: String,
    pub is_open: bool,
}
impl ResourceManager {
    pub fn new(name: String) -> Self {
        Self {
            name,
            is_open: false,
        }
    }
    pub fn __enter__(&mut self) -> &Self {
        self.is_open = true;
        return self;
    }
    pub fn __exit__(
        &mut self,
        exc_type: serde_json::Value,
        exc_val: serde_json::Value,
        exc_tb: serde_json::Value,
    ) -> bool {
        self.is_open = false;
        return false;
    }
    pub fn use_resource(&self) -> i32 {
        if self.is_open {
            return 42;
        };
        return 0;
    }
}
#[derive(Debug, Clone)]
pub struct Counter {
    pub max_count: i32,
    pub count: i32,
}
impl Counter {
    pub fn new(max_count: i32) -> Self {
        Self {
            max_count,
            count: 0,
        }
    }
    pub fn __iter__(&self) -> &Self {
        return self;
    }
    pub fn __next__(&mut self) -> i32 {
        if self.count < self.max_count {
            self.count = self.count + 1;
            return self.count;
        };
        return -1;
    }
}
#[doc = "Demonstrate with statement support"]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn demo_with_statement() {
    let _context = ResourceManager::new("test".to_string().to_string());
    let rm = _context.__enter__();
    let result = rm.use_resource();
    result
}
#[doc = "Demonstrate iterator protocol"]
#[doc = " Depyler: verified panic-free"]
pub fn demo_iterator() -> i32 {
    let mut counter = Counter::new(3);
    let mut total = 0;
    let mut val = counter.__next__();
    while val != -1 {
        total = total + val;
        val = counter.__next__();
    }
    total
}
#[doc = "Run all v1.3.0 feature demos"]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn main() {
    let with_result = demo_with_statement();
    let iter_result = demo_iterator();
    with_result + iter_result
}
#[cfg(test)]
mod tests {
    use super::*;
    use quickcheck::{quickcheck, TestResult};
    #[test]
    fn test_demo_iterator_examples() {
        let _ = demo_iterator();
    }
}
